lib/zk_recipes/cache.rb

```diff
@@ -76,18 +144,27 @@ def setup_callbacks(zk_client)
 
         @logger&.debug("on_connected: new session")
         @pending_updates.clear
-        @registered_values.each do |path, _value|
-          @pending_updates[path] = nil unless update_cache(path)
+        @static_paths.each_key do |path|
+          @pending_updates.put_if_absent(path, :static) unless update_static_path(path)
+        end
+        @directory_paths.each_key do |path|
+          @pending_updates.put_if_absent(path, :directory) unless update_directory_path(path)
+        end
+        @runtime_paths.each_key do |path|
+          @pending_updates.put_if_absent(path, :runtime) unless update_runtime_path(path)
         end
         @session_id = @zk.session_id
         @latch.count_down
+
+        @zk.defer(method(:process_pending_updates))
       end
 
       @zk.on_exception do |e|
         @logger&.error { "on_exception: exception=#{e.inspect} backtrace=#{e.backtrace.inspect}" }
       end
     end
 
+    # This method is not threadsafe. All configuration should be done on a single thread.
```

Comment:
Kinda wonder if we should make a YARDoc macro for threat safety so it shows up in the docs as such.

Reply:
Possible to add custom tags, and in this project that may be desirable: 

https://rubydoc.info/gems/yard/file/docs/Tags.md#Adding_Custom_Tags

The annoying part is having to define a custom tag in the YARD runtime, or finding where you can set those globally:

```
yard --tag threadsafe:"This method is threadsafe"
```